Structure and wording update to Readme (#22)
Dynatrace/fastlane-plugin-dynatrace: README.md

fastlane add_plugin dynatrace
```

> ⚠️  Apple enabled two-factor authentication in a way that interferes with a fully automated workflow. The workaround requires manual interaction see section 'App Store Connect Two-Factor-Authentication' below for details.

Reviewer: ```suggestion
⚠️  The way Apple introduced the two-factor authentication interferes with a fully automated workflow. The workaround requires manual interaction. For more information, see **App Store Connect Two-Factor-Authentication** section below for details.
```

## About the Dynatrace fastlane plugin
The Dynatrace fastlane plugin manages uploading symbol files (iOS) or obfuscation mapping files (Android) to the Dynatrace cluster. Symbol and mapping files are used to make reported stack traces human-readable. The plugin also allows to download the latest dSYM files from App Store Connect, which enables full automation of the mobile app deployment, and the pre-processing of dSYM files, a step that is necessary for the Dynatrace cluster to be able to symbolicate.

The plugin provides a single action `dynatrace_process_symbols`. The configuration depends on whether the app is (A) iOS and Bitcode-enabled or (B) iOS and not Bitcode-enabled or an Android app.

For Bitcode-enabled iOS apps we recommend to let the plugin handle the download of the dSYM files from App Store Connect and upload to Dynatrace.

## Usage
To get started, first, ask your Dynatrace administrator for an API token - it can be generated in the global settings in "Integration > Dynatrace API". The token needs the permission 'Mobile symbolication file management' and is used by the plugin to obtain permission to upload the symbol and mapping files to Dynatrace.

> Add the action `dynatrace_process_symbols` to your Fastfile, see further below for all configuration options and a default configuration.

Reviewer: ```suggestion
Add the action `dynatrace_process_symbols` to your Fastfile. You'll find all the configuration options and a default configuration later in the readme.
```

Now, when you run fastlane, the Dynatrace plugin will manage the symbol files of your app as configured.

### Dynatrace Managed (1.195 and earlier)

> For cluster versions 1.195 and earlier the Dynatrace application 'Symbolication Client' has to be downloaded manually and explicitly specified (`dtxDssClientPath`). For all cluster versions above 1.195 it is fetched and updated automatically. A matching version can be downloaded manually with this link [https://api.mobileagent.downloads.dynatrace.com/sprint-latest-dss-client/xyz](https://api.mobileagent.downloads.dynatrace.com/sprint-latest-dss-client/xyz) by replacing `xyz` with the 3-digit sprint version of your Dynatrace Managed installation.

Reviewer: ```suggestion
For cluster versions 1.195 and earlier, the Dynatrace application 'Symbolication Client' has to be downloaded manually and explicitly specified (`dtxDssClientPath`). For all cluster versions above 1.195 it is fetched and updated automatically. A matching version can be downloaded manually with this link [https://api.mobileagent.downloads.dynatrace.com/sprint-latest-dss-client/xyz](https://api.mobileagent.downloads.dynatrace.com/sprint-latest-dss-client/xyz) by replacing `xyz` with the 3-digit sprint version of your Dynatrace Managed installation.
```

## A) Bitcode-enabled iOS apps

> This is the right option if your app is distributed via Apple's App Store or TestFlight and Bitcode-enabled. For all other cases look at option B below.

Reviewer: ```suggestion
This is the right approach if your app is distributed via Apple's App Store or TestFlight and Bitcode-enabled. For all other cases, follow the approach B below.
```

Background: If your app is Bitcode-enabled, then the dSYMs that are generated during the Xcode build are **not** the dSYMs that need to be uploaded to Dynatrace. Apple recompiles the application on their servers, generating new dSYM files in the process. These newly generated dSYM files need to be downloaded from *App Store Connect*, processed and uploaded to Dynatrace.

### Automatically downloading dSYMs

> To fully automate the following five step workflow, add the snippets below to the respective files and fill in the placeholders. Uploading the app the App Store Connect is a necessary prerequisite and either handled manually or by fastlane directly:

Reviewer: ```suggestion
To fully automate the following five-step workflow, add the snippets below to the respective files and fill in the placeholders. Uploading the app the App Store Connect is a necessary prerequisite and either handled manually or by fastlane directly:
```

fastlane add_plugin dynatrace
```

⚠️  Apple enabled two-factor authentication in a way that interferes with a fully automated workflow. The workaround requires manual interaction see section 'App Store Connect Two-Factor-Authentication' below for details.

## About the Dynatrace fastlane plugin
The Dynatrace fastlane plugin manages uploading symbol files (iOS) or obfuscation mapping files (Android) to the Dynatrace cluster. Symbol and mapping files are used to make reported stack traces human-readable. The plugin also allows to download the latest dSYM files from App Store Connect, which enables full automation of the mobile app deployment, and the pre-processing of dSYM files, a step that is necessary for the Dynatrace cluster to be able to symbolicate.

The plugin provides a single action `dynatrace_process_symbols`. The configuration depends on whether the app is (A) iOS and Bitcode-enabled or (B) iOS and not Bitcode-enabled or an Android app.

> For Bitcode-enabled iOS apps we recommend to let the plugin handle upload of the app to App Store Connect and download of the dSYM files.

Reviewer 2: the plugin only handles dlsym download, processing & upload to Dynatrace, correct? The upload of the app is not part of this and usually done with normal fastlane actions?

Author: Didn't realize that. It's fixed in the new commit.

## Usage

> To get started, first, ask your Dynatrace administrator for a an API token - it can be generated in the global settings in "Integration > Dynatrace API". The token needs the permission 'Mobile symbolication file management' and is used by the plugin to obtain permission to upload the symbol and mapping files to Dynatrace.

Reviewer 3: This only applies to the v1 API, is there a way to do this through v2 already? Maybe this should be mentioned here.

Author: No, v1 API only.

Reviewer 2: "for (a) an API token"

## A) Bitcode-enabled iOS apps
This is the right option if your app is distributed via Apple's App Store or TestFlight and Bitcode-enabled. For all other cases look at option B below.

Background: If your app is Bitcode-enabled, then the dSYMs that are generated during the Xcode build are **not** the dSYMs that need to be uploaded to Dynatrace. Apple recompiles the application on their servers, generating new dSYM files in the process. These newly generated dSYM files need to be downloaded from *App Store Connect*, processed and uploaded to Dynatrace.

### Automatically downloading dSYMs

To fully automate the following five step workflow, add the snippets below to the respective files and fill in the placeholders:

> 1. Upload build to App Store Connect

Reviewer 2: again, the plugin doesn't upload the app - only waits for processing to finish

so the plugin should be called after uploading the app to apple (with normal fastlane action)

	appId: "<Dynatrace application ID>",
	apitoken: "<Dynatrace API token>",
	os: "ios",
	bundleId: "<CFBundlebundleId>",
	versionStr: "<Build Number (CFBundleVersion)>",
	version: "<App Version (CFBundleShortVersionString)>",
	server: "<Dynatrace Environment URL>",
	downloadDsyms: true
)
```

#### Waiting time between app upload and dSYM file download
There is a waiting time after the application has finished uploading to App Store Connect until the dSYM files are ready to be downloaded. The Dynatrace fastlane plugin waits and downloads the symbol files if setting the `waitForDsymProcessing` is true and a waiting period is provided via `waitForDsymProcessingTimeout`. We recommend 1800 seconds (30 mins) as the default waiting time. In our exprience this is sufficiently long for the processing to happen. If this duration is not long enough it needs to be increased. 

> Note: this timeout is the **maximum** waiting time. If the symbol files are ready sooner, the plugin will continue to the download and will not wait for the whole duration of the timeout.

## B) Not Bitcode-enabled iOS apps or Android apps

> If one of the following options is true, then this option is the right one for you:

Reviewer: ```suggestion
If at least one of the following conditions is true, you should follow this approach:
```

| server                       | The API endpoint for the Dynatrace environment (e.g. `https://environmentID.live.dynatrace.com` or `https://dynatrace-managed.com/e/environmentID`).                  |                |
| debugMode                    | Enable debug logging.                                                                                                                                                 | false          |

## App Store Connect Two-Factor-Authentication

> When the plugin is used to download symbols from *App Store Connect* automatically (`downloadDsyms: true`) valid login App Store Connect credentials with access to the dSYM files are required. The preferred method of doing so is by setting the `FASTLANE_USER` and `FASTLANE_PASSWORD` environment variables to their respective values.

Reviewer: ```suggestion
When the plugin is used to download symbols from *App Store Connect* automatically (`downloadDsyms: true`),  valid App Store Connect credentials with access to the dSYM files are required. The preferred method of doing so is by setting the `FASTLANE_USER` and `FASTLANE_PASSWORD` environment variables to their respective values.
```

Apple started enforcing 2-Factor-Authentication for the *App Store Connect* API in February 2021. This [limits the ability to automate the symbol processing](https://github.com/fastlane/fastlane/discussions/17655), because it will most likely involve manual interaction, which is not suitable for CI automation. The only workaround at this point in time is to pre-generate a session and cache it in CI.

### Fastlane Session
The full documentation for this can be found on the [fastlane docs](https://docs.fastlane.tools/best-practices/continuous-integration/#two-step-or-two-factor-auth
) under **spaceauth**.

You can generate a session by running `fastlane spaceauth -u user@email.com` on your machine and copy the output into an environment variable `FASTLANE_SESSION` on the target system (e.g. CI).

> Note: 

> > - Session is only valid for the "region" you created it in. If you CI is in a different geographical location the authentication might fail

Reviewer: ```suggestion
> - Session is only valid for the "region" you created it in. If your CI is in a different geographical location, the authentication might fail
```

| server                       | The API endpoint for the Dynatrace environment (e.g. `https://environmentID.live.dynatrace.com` or `https://dynatrace-managed.com/e/environmentID`).                  |                |
| debugMode                    | Enable debug logging.                                                                                                                                                 | false          |

## App Store Connect Two-Factor-Authentication
When the plugin is used to download symbols from *App Store Connect* automatically (`downloadDsyms: true`) valid login App Store Connect credentials with access to the dSYM files are required. The preferred method of doing so is by setting the `FASTLANE_USER` and `FASTLANE_PASSWORD` environment variables to their respective values.

> Apple announced that 2-Factor-Authentication for the *App Store Connect* API will be enforced starting February 2021. This [limits the ability to automate the symbol processing](https://github.com/fastlane/fastlane/discussions/17655), because it will most likely involve manual interaction, which is not suitable for CI automation. The only workaround at this point in time is to pre-generate a session and cache it in CI.

Reviewer 2: This already happened. The fastlane session is currently the only way to (semi-)automate this, as described below. 

We need to watch out for when dsym APIs with an API Key are available from Apple & fastlane - which will hopefully happen some time soon..
